### PR TITLE
Dynamically calculate number of trajectory elements

### LINKED ISSUE #2128: TrajectoryPlanner should calculate num_elements based on path length
### Description of the task

There is an implicit requirement that our robot controller/firmware runs **_at least_** as fast as the shortest `time_segment` in a `TrajectoryPlanner` planned trajectory.

[Currently for some given path (eg. the `MovePrimitive` moving to a destination), we create `PathParameters` with `num_segments=10`](https://github.com/UBC-Thunderbots/Software/blob/c71c01171ce697d077a3a0f17399eeb3cb195d85/src/firmware/app/primitives/move_primitive.c#L105). This will divide the path into 10 segments. Notably for control purposes, each segment corresponds to a single "control command", eg. accelerate, coast, decelerate. Therefore, shorter paths will have shorter segments, and therefore there will be less time between each segment.

If the firmware/control is running slower than the `TrajectoryPlanner` needs it to, [it will "skip ahead" segments until the current segment matches the current time](https://github.com/UBC-Thunderbots/Software/blob/c71c01171ce697d077a3a0f17399eeb3cb195d85/src/firmware/app/primitives/move_primitive.c#L134). This is so that we don't execute "old" control commands if we are already ahead. However skipping control commands can result in the robots moving very slowly, especially if we skip some "acceleration" or "coast" commands and decelerate instead.

This is what we observed currently in simulation with our custom simulator (on the latest `master`, commit `c71c01171ce697d077a3a0f17399eeb3cb195d85` at the time of writing).

We simulate our firmware at 200Hz. However for certain short motions, especially pivots, the difference in time between segments is sometimes `< 1/200 seconds`. This causes the robots to sometimes pivot really slowly.

To fix this, we need a "stronger" guarantee that we won't end up with time_segments smaller than the max speed we can run control. The ideal fix for this is likely to calculate the number of segments a path/trajectory should have, so that shorter paths get divided into fewer segments, avoiding tiny segments and therefore tiny unachievable timesteps.

**IMPORTANT NOTE: This can ONLY be reproduced with the StandaloneSimulator. It can NOT be reproduced in simulated tests because everything runs in lockstep**

### Acceptance criteria

- [ ] Calculate the num_segments for a trajectory planner path based on the path length

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Dynamically calculate number of trajectory elements based estimated time to complete the trajectory

### Testing Done
manual testing that it results in faster movement

### Resolved Issues
resolves https://github.com/UBC-Thunderbots/Software/issues/2128

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by MathewMacDougall — COMMENTED** [review]
Minor nits regarding division by zero, otherwise looks reasonable. What do you want to do with the "temporary fix" PR once this is merged? Merge it anyway, or drop it? (at the very least we should make the simulation rate the same as CONTROL_LOOP_HZ for parity)

**Comment by MathewMacDougall** [comment]
JK just saw you closed the other one. In that case, can you change the default value of `DEFAULT_PHYSICS_TIME_STEP_SECONDS` in `simulator.h` to be derived from `CONTROL_LOOP_HZ`? That way the sim should be running control at the same rate as the actual firmware.

File: src/firmware/app/primitives/move_primitive.c [hunk 1]
```diff
@@ -90,19 +92,28 @@ void app_move_primitive_start(TbotsProto_MovePrimitive prim_msg, void* void_stat
...
         min_angle_delta(current_orientation, destination_orientation);
+    const float orientation_delta =
+        net_change_in_orientation + (float)revolutions_to_spin * 2.0f * (float)M_PI;
+
+    const float estimated_time_delta = fmaxf(
+        fabsf(distance_to_destination) / (float)(ROBOT_MAX_SPEED_METERS_PER_SECOND),
+        fabsf(net_change_in_orientation) / (float)(ROBOT_MAX_ANG_SPEED_RAD_PER_SECOND));
+
+    // clamp num elements between 3 and TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS
+    const unsigned int num_elements = (unsigned int)fmaxf(
```
Comment (jared-paul) [inline src/firmware/app/primitives/move_primitive.c:104]: what's the significance of 3?

Comment (jonathanlew) [reply src/firmware/app/primitives/move_primitive.c:104]: it's the minimum number of elements, added a comment

File: src/firmware/app/control/control.c [hunk 2]
```diff
@@ -38,22 +38,31 @@ float app_control_getMaximalTorqueScaling(const ForceWheel_t* force_wheels[4],
...
         if (effective_voltage > max_effective_motor_voltage)
         {
             max_effective_motor_voltage = effective_voltage;
         }
-    }
 
-    float emf_ratio_min = battery_voltage / max_effective_motor_voltage;
+        if (fabsf(resistive_voltage_loss) != 0)
+        {
+            float slip_ratio = constants.motor_max_voltage_before_wheel_slip /
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/firmware/app/control/control.c]: oh wow didn't think of that

[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/firmware/app/control/control.c]: adding 1e-6 also runs into overflow problems but it's much less likely

**Review by MathewMacDougall — DISMISSED** [review]
LgucciTM

**Review by jared-paul — DISMISSED** [review]
LGTM!

**Review by jared-paul — APPROVED** [review]

**Review by MathewMacDougall — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**